chromium_src/chrome/browser/ui/startup/startup_browser_creator.cc

```diff
@@ -0,0 +1,66 @@
+/* Copyright (c) 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/tor/tor_profile_manager.h"
+#include "brave/browser/tor/tor_profile_service_factory.h"
+#include "brave/common/brave_switches.h"
+#include "brave/components/tor/tor_profile_service.h"
+#include "chrome/browser/ui/startup/startup_browser_creator_impl.h"
+
+class BraveStartupBrowserCreatorImpl final : public StartupBrowserCreatorImpl {
+ public:
+  BraveStartupBrowserCreatorImpl(const base::FilePath& cur_dir,
+                                 const base::CommandLine& command_line,
+                                 chrome::startup::IsFirstRun is_first_run);
+
+  BraveStartupBrowserCreatorImpl(const base::FilePath& cur_dir,
+                                 const base::CommandLine& command_line,
+                                 StartupBrowserCreator* browser_creator,
+                                 chrome::startup::IsFirstRun is_first_run);
+
+  bool Launch(Profile* profile,
+              const std::vector<GURL>& urls_to_open,
+              bool process_startup,
+              std::unique_ptr<LaunchModeRecorder> launch_mode_recorder);
+};
+
+BraveStartupBrowserCreatorImpl::BraveStartupBrowserCreatorImpl(
+    const base::FilePath& cur_dir,
+    const base::CommandLine& command_line,
+    chrome::startup::IsFirstRun is_first_run)
+    : StartupBrowserCreatorImpl(cur_dir, command_line, is_first_run) {}
+
+BraveStartupBrowserCreatorImpl::BraveStartupBrowserCreatorImpl(
+    const base::FilePath& cur_dir,
+    const base::CommandLine& command_line,
+    StartupBrowserCreator* browser_creator,
+    chrome::startup::IsFirstRun is_first_run)
+    : StartupBrowserCreatorImpl(cur_dir,
+                                command_line,
+                                browser_creator,
+                                is_first_run) {}
+
+// If the --tor command line flag was provided, switch the profile to Tor mode
+// and then call the original Launch method.
+bool BraveStartupBrowserCreatorImpl::Launch(
+    Profile* profile,
+    const std::vector<GURL>& urls_to_open,
+    bool process_startup,
+    std::unique_ptr<LaunchModeRecorder> launch_mode_recorder) {
+  if (StartupBrowserCreatorImpl::command_line_.HasSwitch(switches::kTor)) {
+    LOG(INFO) << "Switching to Tor profile and starting Tor service.";
+    profile = TorProfileManager::GetInstance().GetTorProfile(profile);
+    tor::TorProfileService* service =
+        TorProfileServiceFactory::GetForContext(profile);
+    service->RegisterTorClientUpdater();
```

Comment:
this seems sketchy to me. @darkdh is there someplace else we can put this in TorProfileManager so it gets called just like in `SwitchToTorProfile`?

Reply:
yea, I think we can do that in ` TorProfileManager::GetTorProfile`, right before returning the Tor profile. Since Tor profile is useless if there is no Tor binary to be launched.
Also as discussed in DM, we should do `GetOriginalProfile()` before `GetOffTheRecordProfile()` so even if OTR profile get passed in, we can still handle it.

Reply:
The latter should be addressed in 590c5f85b60d2a9c4de6829f6372b3c0e181d3ab.

Reply:
...and the former is addressed in e601a077fae843b1f4858bcc5380150d5a63d0cb.